stream/stream.go

```diff
@@ -0,0 +1,194 @@
+				// returned. This path serves as "what are you doing
+				// here? Go away" in the window between the call
+				// returning and the handler registration being
+				// cleaned up.
+				return nil, ctx.Err()
+			case <-callbackCtx.Done():
+				// Server cancellation (or maybe indirectly client
+				// cancellation too), we'll be told why when our Call
+				// below returns.
+				return nil, callbackCtx.Err()
+			}
+		})
+
+		errch := make(chan error, 1)
+		go func() {
+			// Note, unregister the capability handler in this
+			// goroutine, not the calling context, so that the other
+			// side can't hit "method not found" as the iterator shuts
+			// down.
+			defer peer.Handle(capability, nil)
+			defer close(errch)
+			_, err := peer.Call(ctx, method, req)
+			if ctx.Err() != nil {
+				// This codepath isn't strictly necessary, but it
+				// makes tests deterministic: a client-side
+				// cancellation can manifest in two ways: the
+				// peer.Call() notices first and returns ctx.Err()
+				// locally; or a running stream callback sends a
+				// cancellation error to the server, which then
+				// bounces it back to us, and peer.Call returns that.
+				//
+				// This makes it tricky to reliably test client-side
+				// cancellation, because the error is either a simple
+				// local "context canceled", or it's a "context
+				// canceled" wrapped in two layers of chirp.CallError,
+				// with type information stripped off.
+				//
+				// This codepath prioritizes client cancellation over
+				// whatever peer.Call returns, so that regardless of
+				// how cancellation manifests, we reliably report it
+				// as a local cancellation.
+				//
+				// Plus, aside from making unit tests deterministic,
+				// it's less confusing to report a locally cancelled
+				// context as a local error, rather than as "we called
+				// the peer, it called something else, and that other
+				// thing returned an error".
+				errch <- ctx.Err()
+			} else {
+				errch <- err
+			}
+		}()
+
+		for {
+			select {
+			case v := <-vals:
+				if !yield(v, nil) {
+					// Returning cancels the context that both the
+					// peer call and callback run in, so they'll
+					// gracefully unwind and clean themselves up. We
+					// don't need to wait for that to happen.
+					return
+				}
+			case err := <-errch:
+				if err != nil {
+					yield(nil, err)
+				}
+				return
+			case <-ctx.Done():
+				yield(nil, ctx.Err())
+				return
+			}
+		}
+	}
+}
+
+// HandlerFunc is a variant of chirp.Handler that yields a stream of
+// responses, rather than a single value. The returned iterator is
+// expected to only yield a non-nil error as its final element,
+// following zero or more error-free tuples.
+type HandlerFunc func(context.Context, *chirp.Request) iter.Seq2[[]byte, error]
+
+// Handle adapts a HandlerFunc into a chirp.Handler. The resulting
+// handler must be invoked with [Call].
+func Handle(peer *chirp.Peer, method string, fn HandlerFunc) {
```

Comment:
I suppose we don't really even need to deal with the registration here, if we don't want—this could be `Handler(method string, fn HandlerFunc) chirp.Handler` and the caller can plug it in when they're ready. (But this is also fine).